### PR TITLE
RDD Interpreter

### PR DESCRIPTION
**Pull request by jbouffard**
This PR adds the ability to use `RDD[K, Tile] with Metadata[TileLayerMetadata[K]]` in Maml.

### CODE REVIEW

File: rdd/src/main/scala/ast/RDDFoldableExpressions.scala [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+import maml.ast._
+
+import cats.data._
+import Validated._
+
+import java.security.InvalidParameterException
+
+
+trait RDDFoldableExpression extends Expression {
```
Comment (echeipesh) [inline rdd/src/main/scala/ast/RDDFoldableExpressions.scala]: This is the same type as `FoldableExpression` is it really needed?

File: rdd/src/main/scala/ast/RDDFoldableExpressions.scala [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+      case (MamlKind.Double, RDDMamlKind.SpaceTimeRDD) => RDDMamlKind.SpaceTimeRDD
+      case (RDDMamlKind.SpaceTimeRDD, MamlKind.Double) => RDDMamlKind.SpaceTimeRDD
+      case (RDDMamlKind.SpaceTimeRDD, RDDMamlKind.SpaceTimeRDD) => RDDMamlKind.SpaceTimeRDD
+
+      case (x1, x2) => throw new InvalidParameterException(s"Expected rdd, int, or double. Found $x1 $x2")
+    }
+}
+
+case class RDDAddition(children: List[Expression]) extends Operation with RDDFoldableExpression {
+  val kindDerivation = RDDFoldableExpression.rddOrScalarDerivation(this)(_, _)
```
Comment (echeipesh) [inline rdd/src/main/scala/ast/RDDFoldableExpressions.scala]: Ah, I see, you need something to augment the kind derivation, to possible return RDD result

**Comment by moradology** [comment]
I think this needs a rebase with master's `.travis` changes

**Comment by jbouffard** [comment]
@echeipesh This has been updated and is good to review.

File: rdd/src/main/scala/ast/RDDSource.scala [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+
+case class SpatialRDDLiteral(value: TileLayerRDD[SpatialKey]) extends Source {
+  val id = value.toString
+  val kind = MamlKind.Tile
+}
+
+case class SpatialRDDSource(id: String) extends UnboundSource {
+  val kind = MamlKind.Tile
+}
```
Comment (echeipesh) [inline rdd/src/main/scala/ast/RDDSource.scala:15]: @moradology Do you think we need this case class or can we use `TileSource` in all cases where we have a variable of `MamlKind.Tile` ?

File: rdd/src/main/scala/eval/RDDResult.scala [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+import org.apache.spark.rdd._
+
+import scala.reflect.ClassTag
+
+
+case class SpatialRDDResult(res: TileLayerRDD[SpatialKey]) extends Result {
+  def as[T](implicit ct: ClassTag[T]): Interpreted[T] = {
+    val cls = ct.runtimeClass
+    if (classOf[RDD[(SpatialKey, Tile) with Metadata[TileLayerMetadata[SpatialKey]]]] isAssignableFrom cls)
```
Comment (echeipesh) [inline rdd/src/main/scala/eval/RDDResult.scala]: Should be `classOf[RDD[(SpatialKey, Tile)] with Metadata[TileLayerMetadata[SpatialKey]]]`

File: rdd/src/main/scala/eval/directive/RDDOpDirectives.scala [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+            rdd1.withContext { _ localAdd rdd2 }
+        })
+      SpatialRDDResult(rddSums.withContext { _ localAdd sums })
+      })
+  }
+
+  val subtractionDirective = Directive { case (a@Subtraction(_), childResults) =>
+    val results = childResults.reduce({ (res1: Result, res2: Result) =>
+      (res1, res2) match {
+        case (SpatialRDDResult(rdd1), SpatialRDDResult(rdd2)) => SpatialRDDResult(rdd1.withContext{ _ - rdd2 })
```
Comment (echeipesh) [inline rdd/src/main/scala/eval/directive/RDDOpDirectives.scala]: what if `rdd1` and `rdd2` are not aligned, their `LayoutDefinition` is not matched or they are in difference CRS? There should be an error based on that.

File: rdd/src/main/scala/ast/codec/RDDMamlUtilityCodecs.scala [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+import io.circe.parser._
+import io.circe.optics.JsonPath._
+import cats.syntax.functor._
+
+import geotrellis.spark._
+
+
+trait RDDMamlUtilityCodecs {
+  implicit val rddMamlKindDecoder: Decoder[MamlKind] = Decoder[String].emap({
+    case "spatialRDD" => Right(MamlKind.Tile)
```
Comment (echeipesh) [inline rdd/src/main/scala/ast/codec/RDDMamlUtilityCodecs.scala]: It doesn't seem like these codecs can actually decode a full Expression tree, just the RDD literal. I would suggest making decoding of this a follow up PR where we need to create an explicit pattern how to extend the decoders. @moradology confirm/deny?

**Pull request merged**